src/librustc/middle/resolve_lifetime.rs

```diff
@@ -2121,6 +2121,10 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
             _ => false,
         };
 
+        // Only skip `Self`, `&Self` and `&mut Self`,
+        // and to handle other `self`s like normal arguments.
+        let mut skip = 0;
```

Comment:
Would be better as a `bool` and then cast to `usize` as needed. In so doing, also rename to `should_skip` or something like that.